Mathlib/Algebra/Star/StarRingHom.lean

```diff
@@ -0,0 +1,469 @@
+/-
+Copyright (c) 2024 Christopher Hoskin. All rights reserved.
+Released under Apache 2.0 license as described in the file LICENSE.
+Authors: Christopher Hoskin
+-/
+import Mathlib.Algebra.Ring.Defs
+import Mathlib.Algebra.Star.Basic
+
+/-!
+# Morphisms of star rings
+
+This file defines morphisms between (non-unital) rings `A` and `B` where both
+`A` and `B` are equipped with a `star` operation. This morphisms, namely `NonUnitalStarRingHom` is
+a direct extension of its non-`star`red counterpart with a field `map_star` which guarantees it
+preserves the star operation.
+
+As with `NonUnitalRingHom`, the multiplications are not assumed to be associative or unital.
+
+## Main definitions
+
+  * `NonUnitalStarRingHom`
+
+## Implementation
+
+This file is heavily inspired by `Mathlib.Algebra.Star.StarAlgHom`.
+
+## Tags
+
+non-unital, ring, morphism, star
+-/
+
+open EquivLike
+
+/-! ### Non-unital star ring homomorphisms -/
+
+/-- A *non-unital ⋆-ring homomorphism* is a non-unital ring homomorphism between non-unital
+non-associative semirings `A` and `B` equipped with a `star` operation, and this homomorphism is
+also `star`-preserving. -/
+structure NonUnitalStarRingHom (A B : Type*) [NonUnitalNonAssocSemiring A]
+  [Star A] [NonUnitalNonAssocSemiring B]
+  [Star B] extends A →ₙ+* B where
+  /-- By definition, a non-unital ⋆-ring homomorphism preserves the `star` operation. -/
+  map_star' : ∀ a : A, toFun (star a) = star (toFun a)
+
+/-- `α →⋆ₙ+* β` denotes the type of non-unital ring homomorphisms from `α` to `β`. -/
+infixr:25 " →⋆ₙ+* " => NonUnitalStarRingHom
+
+/-- Reinterpret a non-unital star ring homomorphism as a non-unital ring homomorphism
+by forgetting the interaction with the star operation. -/
+add_decl_doc NonUnitalStarRingHom.toNonUnitalRingHom
+
+/-- `NonUnitalStarRingHomClass F A B` states that `F` is a type of non-unital ⋆-ring homomorphisms.
+You should also extend this typeclass when you extend `NonUnitalStarRingHom`. -/
+class NonUnitalStarRingHomClass (F : Type*) (A B : outParam Type*)
+     [NonUnitalNonAssocSemiring A] [Star A] [NonUnitalNonAssocSemiring B] [Star B]
+    [FunLike F A B] [NonUnitalRingHomClass F A B] extends StarHomClass F A B : Prop
+
+namespace NonUnitalStarRingHomClass
+
+variable {F A B : Type*}
+variable [NonUnitalNonAssocSemiring A] [Star A]
+variable [NonUnitalNonAssocSemiring B] [Star B]
+variable [FunLike F A B] [NonUnitalRingHomClass F A B]
+
+/-- Turn an element of a type `F` satisfying `NonUnitalStarRingHomClass F A B` into an actual
+`NonUnitalStarRingHom`. This is declared as the default coercion from `F` to `A →⋆ₙ+ B`. -/
+@[coe]
+def toNonUnitalStarRingHom [NonUnitalStarRingHomClass F A B] (f : F) : A →⋆ₙ+* B :=
+  { (f : A →ₙ+* B) with
+    map_star' := map_star f }
+
+instance [NonUnitalStarRingHomClass F A B] : CoeTC F (A →⋆ₙ+* B) :=
```

Comment:
I *think* this should be a `CoeHead` instance. If Lean doesn't complain, let's go with that.
```suggestion
instance [NonUnitalStarRingHomClass F A B] : CoeHead F (A →⋆ₙ+* B) :=
```